### PR TITLE
Support ES6 syntax

### PR DESCRIPTION
**Pull request by tricknotes**
## Abstract

We can use es6 syntax on asset pipeline.
This way is the same as [ember-cli](https://github.com/stefanpenner/ember-cli) .

The setting specified ember-rails is the followings:

* `my-app.js.es6`
  ``` javascript
  //= require ember-rails/application
  //= require_tree
  //= require_self

  import App from 'ember-rails/application';

  App.create();
  ```

* `application.js`
  ``` javascript
  // = require ./my-app

  require('my-app');
  ```

I think the following benefits are there for ember-rails users:
* It supports ES6 (current Ember way) in ember-rails.
* It helps migrating from ember-rails ls to Ember CLI

## TODO

* [x] Support es6 module syntax
* [x] Support generator option `javascript-engine=es6`
* [x] Add test
* [x] Write how to use es6 module syntax to README
* [x] Write about how to use multiple app
* [x] Change ES6 transpiler

### CODE REVIEW

**Comment by ashrafmajdee** [comment]
+1

**Comment by bradleypriest** [comment]
Awesome, do you think we'll be able to support `.coffee.es6` or `.js.es6.coffee` at some point as well?

**Comment by tricknotes** [comment]
Yes, It is important but now it doesn't be supported.
I'll support it at some early date.

**Comment by mmun** [comment]
Looking forward to seeing this land! :+1:

File: README.md [hunk 1]
```diff
@@ -94,15 +94,46 @@ You can now use the flag `--javascript-engine=em` to specify EmberScript
...
+`require ember-rails/application` is important.
+It provides customized Ember application to resolve dependencies from ES6 modules.
+
+`application.js`
+``` javascript
+//= require jquery
+//= require handlebars
+//= require ember
+//= require ember-data
+//= require ./my_app
```
Comment (stefanpenner) [inline README.md]: lets try to align this with ember-cli's conventions, all files/folders in the app are dasherized

File: ember-rails.gemspec [hunk 2]
```diff
@@ -15,6 +15,7 @@ Gem::Specification.new do |s|
   s.add_dependency "railties", [">= 3.1"]
   s.add_dependency "active_model_serializers"
   s.add_dependency "barber", [">= 0.6.0"]
+  s.add_dependency "es6_module_transpiler-rails"
```
Comment (stefanpenner) [inline ember-rails.gemspec]: we hope to move ti esperanto for the module transpiles soon https://github.com/ef4/broccoli-es6modules/pull/4 at somepoint (when we switch) we should coordinate that this switches aswell.

File: lib/ember/rails/version.rb [hunk 3]
```diff
@@ -1,5 +1,9 @@
 module Ember
   module Rails
     VERSION = '0.16.0'
+
+    EMBER_LOADER_VERSION = '2.1.1'
```
Comment (stefanpenner) [inline lib/ember/rails/version.rb]: we should likely stick to the 1x version of the loader.js

File: lib/generators/ember/component_generator.rb [hunk 4]
```diff
@@ -15,7 +15,7 @@ class ComponentGenerator < ::Rails::Generators::NamedBase
 
       def create_component_files
         dashed_file_name = file_name.gsub(/_/, '-')
```
Comment (stefanpenner) [inline lib/generators/ember/component_generator.rb]: :+1:

File: lib/generators/templates/app.js.es6 [hunk 5]
```diff
@@ -0,0 +1,19 @@
+//= require ember-rails/application
+//
```
Comment (stefanpenner) [inline lib/generators/templates/app.js.es6]: should we move the app into a specific folder and then just require the whole tree?

Comment (stefanpenner) [inline lib/generators/templates/app.js.es6]: i know sometimes people have more then 1 ember app in a rails app

File: lib/generators/templates/application_adapter.js.es6 [hunk 6]
```diff
@@ -0,0 +1,11 @@
+// Override the default adapter with the `DS.ActiveModelAdapter` which
+// is built to work nicely with the ActiveModel::Serializers gem.
+export default DS.ActiveModelAdapter.extend({
+  // Adds X-CSRF-Token to all REST requests.
+  // Allows for the use of Rails protect_from_forgery
+  // The CSRF Token is normally found in app/views/layouts/application.html.*
+  // inserted with the rails helper: "csrf_meta_tags"
+  headers: {
+    "X-CSRF-Token": $('meta[name="csrf-token"]').attr('content')
```
Comment (stefanpenner) [inline lib/generators/templates/application_adapter.js.es6]: in ember-cli we have a config module we can import information from, maybe we should consider that?

```js
import config from 'app-name/config/environment';


```

File: README.md (hunk 1 above)
Comment (tricknotes) [inline README.md]: agreed :wink:

File: ember-rails.gemspec (hunk 2 above)
Comment (tricknotes) [inline ember-rails.gemspec]: ember-cli uses [esperanto](https://github.com/ember-cli/ember-cli/blob/v0.1.11/package.json#L90).
On the other hand, blueprints uses [6to5](https://github.com/ember-cli/ember-cli/blob/v0.1.11/blueprints/app/files/package.json#L25).

What is the difference of both of them?

File: lib/generators/templates/app.js.es6 (hunk 5 above)
Comment (tricknotes) [inline lib/generators/templates/app.js.es6]: That's right :+1:

File: ember-rails.gemspec (hunk 2 above)
Comment (rwjblue) [inline ember-rails.gemspec]: @tricknotes - esperanto is just for module transpilation, but 6to5 supports a large number of other ES6 (ES2015?) features.

Comment (tricknotes) [inline ember-rails.gemspec]: Ah, thanks @rwjblue 

In sprockets 4, 6to5 will become  default ES6 transpiler:  https://github.com/sstephenson/sprockets/pull/682 
Should we also use 6to5?

**Comment by tricknotes** [comment]
I'm now working on [another branch](https://github.com/tricknotes/ember-rails/tree/new-es6-module-syntax) because I have to consider implementation carefully.
When it is decided, I'll update this PR.

**Comment by tricknotes** [comment]
Updated to add the following to this PR's description:

> I think the following benefits are there for ember-rails users:
* It supports ES6 (current Ember way) in ember-rails.
* It helps migrating from ember-rails ls to Ember CLI

**Comment by tricknotes** [comment]
All things are done.

@stefanpenner @rwjblue 
Could you review again?

**Comment by tulios** [comment]
Hi @tricknotes, awesome work! Do you have a fork with this change? I would like to try, this definitely will help to migrate to ember-cli.

**Comment by mufid** [comment]
Any update for this?

**Comment by ashrafmajdee** [comment]
@tricknotes  whats the status on this?

**Comment by ashrafmajdee** [comment]
@stefanpenner @rwjblue

**Comment by tricknotes** [comment]
Thanks for your ping, @ashrafmajdee .

I think this is not fully compatible with Ember CLI, however it is enough to be a migration path from ember-rails to Ember CLI.

@stefanpenner @rwjblue 
Are there any worry?

**Comment by rwjblue** [comment]
This looks good to me.

**Comment by tricknotes** [comment]
Thanks @rwjblue !
I'll release it tonight :stars:

File: README.md [hunk 7]
```diff
@@ -94,6 +94,84 @@ You can now use the flag `--javascript-engine=em` to specify EmberScript
...
+
+`my-app.es6`
+``` javascript
+import Application from 'ember-rails/application';
+
+const App = Application.extend({
+  // Configure your application.
+});
+
+App.craete();
```
Comment (mufid) [inline README.md]: Typo?

File: README.md [hunk 8]
```diff
@@ -94,6 +94,84 @@ You can now use the flag `--javascript-engine=em` to specify EmberScript
...
+import Application from 'ember-rails/application';
+import loadInitializers from 'ember/load-initializers';
+
+const App = Application.extend({
+  modulePrefix: 'my-app' // This value should be the same as directory name.
+});
+
+loadInitializers(App, 'my-app');
+
+App.craete();
```
Comment (mufid) [inline README.md]: Typo?

Comment (tricknotes) [inline README.md]: Thanks @mufid !
Fixed by https://github.com/emberjs/ember-rails/commit/051d326f8a9cad96d7b33a0dff52510a18d0b4d3

**Pull request merged**